core/src/main/java/org/elasticsearch/http/netty/RequestSizeLimitHandler.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.http.netty;
+
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.ESLoggerFactory;
+import org.elasticsearch.http.netty.pipelining.OrderedDownstreamChannelEvent;
+import org.elasticsearch.common.limit.Limiter;
+import org.elasticsearch.common.limit.Reservation;
+import org.jboss.netty.buffer.ChannelBuffer;
+import org.jboss.netty.channel.ChannelDownstreamHandler;
+import org.jboss.netty.channel.ChannelEvent;
+import org.jboss.netty.channel.ChannelHandler;
+import org.jboss.netty.channel.ChannelHandlerContext;
+import org.jboss.netty.channel.ChannelUpstreamHandler;
+import org.jboss.netty.channel.MessageEvent;
+import org.jboss.netty.handler.codec.http.HttpMessage;
+import org.jboss.netty.util.internal.ConcurrentHashMap;
+
+import java.util.Map;
+
+/**
+ * Limits the total size of all HTTP requests that are in flight. On limit breach a <code>CircuitBreakingException</code> is thrown
+ * which should result in HTTP 413 (Request Entity Too Large).
+ */
+@ChannelHandler.Sharable
+public class RequestSizeLimitHandler implements ChannelUpstreamHandler, ChannelDownstreamHandler {
+    private static final ESLogger logger = ESLoggerFactory.getLogger(RequestSizeLimitHandler.class.getName());
+
+    private final Limiter limiter;
+
+    private final Map<Integer, Reservation> reservationsPerChannel = new ConcurrentHashMap<>();
+
+    public RequestSizeLimitHandler(Limiter limiter) {
+        this.limiter = limiter;
+    }
+
+    @Override
+    public void handleUpstream(ChannelHandlerContext ctx, ChannelEvent e) throws Exception {
+        if (e instanceof MessageEvent) {
+            Object message = ((MessageEvent) e).getMessage();
+            if (message instanceof HttpMessage) {
+                ChannelBuffer buf = ((HttpMessage) message).getContent();
+                int channelId = ctx.getChannel().getId();
+                int reservation = buf.readableBytes();
+                // on some occasions, we will get 0 (e.g. when the message is chunked and we get the first chunk)
```

Comment:
The comment here seems wrong,  this handler is after the HTTPChunkHandler instance, so it never gets chunked message as "HttpMessage" here.  But indeed the "HttpMessage" may have empty body,  for example,  HTTP GET requests.